pkg/knownhosts/knownhosts.go

```diff
@@ -0,0 +1,350 @@
+package knownhosts
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/golang.org/x/crypto/ssh"
+)
+
+type KnownHosts struct {
+	Lines []*Line
+}
+
+func Unmarshal(in io.Reader) (*KnownHosts, error) {
+	k := &KnownHosts{
+		Lines: []*Line{},
+	}
+	s := bufio.NewScanner(in)
+	for s.Scan() {
+		l, err := parseHosts(s.Text())
+		if err == nil && l != nil {
+			k.Lines = append(k.Lines, l)
+		} else {
+			fmt.Println(err)
+		}
+	}
+	return k, s.Err()
+}
+
+func (k *KnownHosts) Marshal(out io.Writer) error {
+	for _, l := range k.Lines {
+		b, err := l.Marshal()
+		if err != nil {
+			return err
+		}
+		if _, err := out.Write(b); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+var HostNotFoundError = errors.New("Host is unknown")
+var HostRevokedError = errors.New("Host is revoked")
+var UnsupportedAddrType = errors.New("remote must be a *net.TCPAddr")
+
+func (k *KnownHosts) HostKeyCallback(hostname string, remote net.Addr, key ssh.PublicKey) error {
+	var addr *net.TCPAddr
+	if v, ok := remote.(*net.TCPAddr); ok {
+		addr = v
+	} else {
+		return UnsupportedAddrType
+	}
+	keyBytes := key.Marshal()
+	var matched = []*Host{}
+	for _, l := range k.Lines {
+		if l.CertAuthority {
+			continue
+		}
+		for _, h := range l.Hosts {
+			if h.Match(hostname, addr) && bytes.Equal(keyBytes, l.PublicKey.Marshal()) {
+				if l.Revoked {
+					return HostRevokedError
+				}
+				matched = append(matched, h)
+			}
+		}
+	}
+	if len(matched) == 0 {
+		return HostNotFoundError
+	}
+	return nil
+}
+
+func (k *KnownHosts) AddHost(hostname string, key ssh.PublicKey) error {
+	_parts := strings.Split(hostname, ":")
```

Comment:
Use `net.SplitHostPort`, this could use some tests for IPv6. Also, move this into a `NewLine` function or similar, as we'll probably want to just append a line to the file without rewriting it.